File: ADMIN/Example-API-Flutter/Example.UserService.WebApp/Services/ExamSessionService.cs
using Example.API.Client.Configuration;
using Example.Common.Enums;
using Example.UserService.API.Models;
using Example.UserService.WebApp.Helpers;
using Example.UserService.WebApp.Models.ExamSession;
using Example.UserService.WebApp.Repository.IRepository;
using Example.UserService.WebApp.Services.IServices;
using Microsoft.Extensions.Options;

namespace Example.UserService.WebApp.Services
{
    /// <summary>
    /// Dịch vụ phiên thi: danh sách đang làm/lịch sử, bắt đầu/tiếp tục bài, lưu tiến độ, nộp bài.
    /// Chứa quy tắc phân quyền (chỉ owner), kiểm tra trạng thái session và chuẩn hóa đáp án trước khi gọi API.
    /// </summary>
    public class ExamSessionService : IExamSessionService
    {
        private readonly IExamSessionRepository _sessionRepository;
        private readonly IExamSetRepository _examSetRepository;
        private readonly string _apiBaseUrl;

        // Tiêm repository và cấu hình API base URL (dùng cho màn làm bài / upload ảnh).
        public ExamSessionService(
            IExamSessionRepository sessionRepository,
            IExamSetRepository examSetRepository,
            IOptions<ApiClientOptions> apiOptions)
        {
            _sessionRepository = sessionRepository;
            _examSetRepository = examSetRepository;
            _apiBaseUrl = apiOptions.Value.BaseUrl ?? string.Empty;
        }

        /// <summary>Danh sách phiên thi đang InProgress của user (phân trang 10 bản ghi/trang).</summary>
        public async Task<(List<ExamSessionModel> Items, long TotalItems, string? ErrorMessage)> GetInProgressListAsync(long userId, int pageIndex)
        {
            var result = await _sessionRepository.GetListAsync(new ExamSessionSearchModel
            {
                PageIndex = pageIndex,
                PageSize = 10,
                UserId = userId,
                Status = (int)ExamSessionStatusEnum.InProgress
            });
            return ApiListHelper.ParseList(result);
        }

        /// <summary>Lịch sử thi đã hoàn thành/hết hạn (endpoint GetHistory trên API).</summary>
        public async Task<(List<ExamSessionModel> Items, long TotalItems, string? ErrorMessage)> GetHistoryListAsync(long userId, int pageIndex)
        {
            var result = await _sessionRepository.GetHistoryAsync(new ExamSessionSearchModel
            {
                PageIndex = pageIndex,
                PageSize = 10,
                UserId = userId
            });
            return ApiListHelper.ParseList(result);
        }

        /// <summary>
        /// Lấy phiên thi theo Id và xác minh quyền sở hữu — userId > 0 phải khớp session.UserId.
        /// </summary>
        public async Task<(ExamSessionModel? Session, string? ErrorMessage)> GetOwnedSessionAsync(long id, long userId)
        {
            var result = await _sessionRepository.GetByIdAsync(id);
            if (!result.Success || result.Data == null)
                return (null, result.Message ?? "Không tìm thấy phiên thi.");

            // Quy tắc nghiệp vụ: chặn truy cập phiên thi của người khác.
            if (userId > 0 && result.Data.UserId != userId)
                return (null, "Bạn không có quyền truy cập phiên thi này.");

            return (result.Data, null);
        }

        /// <summary>
        /// Bắt đầu hoặc tiếp tục bài thi: kiểm tra bộ đề (nếu không qua phân công), tìm session InProgress cũ,
        /// rồi gọi API StartExam. Trả ResumeMessage khi API trả về cùng session đang làm.
        /// </summary>
        public async Task<StartExamServiceResult> StartExamAsync(long examSetId, int examType, long examPeriodAssignmentId, long userId)
        {
            var redirectExamSetId = examSetId;
            // Quy tắc nghiệp vụ: thi tự do (không có phân công) phải xác minh bộ đề tồn tại và Active.
            if (examPeriodAssignmentId <= 0)
            {
                var examSetResult = await _examSetRepository.GetByIdAsync(examSetId);
                if (!examSetResult.Success || examSetResult.Data == null || examSetResult.Data.Status != (int)StatusEnum.Active)
                {
                    return new StartExamServiceResult
                    {
                        Success = false,
                        ErrorMessage = "Bộ đề không khả dụng.",
                        RedirectExamSetId = examSetId,
                        ExamPeriodAssignmentId = examPeriodAssignmentId
                    };
                }
            }

            long? existingSessionId = null;
            // Kiểm tra đã có bài đang làm cùng bộ đề + loại thi chưa (để hiển thị thông báo tiếp tục).
            var existingResult = await _sessionRepository.GetListAsync(new ExamSessionSearchModel
            {
                PageIndex = 1,
                PageSize = 1,
                UserId = userId,
                ExamSetId = examSetId > 0 ? examSetId : null,
                ExamType = examType,
                Status = (int)ExamSessionStatusEnum.InProgress
            });
            existingSessionId = existingResult.Data?.FirstOrDefault()?.Id;

            var result = await _sessionRepository.StartExamAsync(new StartExamModel
            {
                ExamPeriodAssignmentId = examPeriodAssignmentId,
                ExamSetId = examSetId,
                ExamType = examType
            });

            if (!result.Success || result.Data == null)
            {
                return new StartExamServiceResult
                {
                    Success = false,
                    ErrorMessage = result.Message ?? "Không thể bắt đầu bài thi.",
                    RedirectExamSetId = redirectExamSetId,
                    ExamPeriodAssignmentId = examPeriodAssignmentId
                };
            }

            return new StartExamServiceResult
            {
                Success = true,
                SessionId = result.Data.Id,
                // Thông báo tiếp tục khi API trả về đúng session InProgress đã có từ trước.
                ResumeMessage = existingSessionId.HasValue && existingSessionId.Value == result.Data.Id
                    ? "Tiếp tục bài thi đang làm."
                    : null,
                RedirectExamSetId = redirectExamSetId,
                ExamPeriodAssignmentId = examPeriodAssignmentId
            };
        }

        /// <summary>Lưu đáp án tạm: chỉ cho phép khi session InProgress và thuộc user.</summary>
        public async Task<(bool Success, string? Message)> SaveProgressAsync(SubmitExamFormModel model, long userId)
        {
            // Lưu đáp án tạm lên API, không chấm điểm.
            if (model.ExamSessionId <= 0)
                return (false, "Phiên thi không hợp lệ.");

            var (session, error) = await GetOwnedSessionAsync(model.ExamSessionId, userId);
            if (session == null)
                return (false, error ?? "Phiên thi không hợp lệ.");
            // Quy tắc nghiệp vụ: không lưu tiến độ khi session đã kết thúc.
            if (session.Status != (int)ExamSessionStatusEnum.InProgress)
                return (false, "Phiên thi đã kết thúc.");

            var result = await _sessionRepository.SaveExamProgressAsync(new SubmitExamModel
            {
                ExamSessionId = model.ExamSessionId,
                Answers = BuildSubmitAnswers(model)
            });

            return (result.Success, result.Message);
        }

        /// <summary>
        /// Chuẩn bị màn làm bài: kiểm tra quyền/trạng thái, tính thời gian kết thúc cho đếm ngược JS.
        /// </summary>
        public async Task<(TakeExamViewModel? ViewModel, string? SuccessMessage, string? ErrorMessage, bool RedirectToDetails, long SessionId)> GetTakeExamAsync(long id, long userId)
        {
            var (session, error) = await GetOwnedSessionAsync(id, userId);
            if (session == null)
                return (null, null, error, false, id);

            // Session đã hết hạn, đã nộp hoặc bị hủy — chuyển về trang chi tiết thay vì màn làm bài.
            if (session.Status != (int)ExamSessionStatusEnum.InProgress)
            {
                var successMessage = session.Status switch
                {
                    (int)ExamSessionStatusEnum.Expired => "Bài thi đã hết giờ và được tính điểm tự động.",
                    (int)ExamSessionStatusEnum.Cancelled => "Bài thi đã bị hủy do vi phạm quy chế thi.",
                    _ => null
                };
                return (null, successMessage, null, true, id);
            }

            var examSetResult = await _examSetRepository.GetByIdAsync(session.ExamSetId);
            var duration = examSetResult.Data?.DurationMinutes ?? 60;
            var startedAt = session.StartedAt == default ? DateTimeOffset.Now : session.StartedAt;

            // EndTimeUnixMs dùng cho đếm ngược JavaScript trên màn làm bài.
            return (new TakeExamViewModel
            {
                Session = session,
                DurationMinutes = duration,
                EndTimeUnixMs = startedAt.AddMinutes(duration).ToUnixTimeMilliseconds(),
                ApiBaseUrl = _apiBaseUrl
            }, null, null, false, id);
        }

        /// <summary>
        /// Nộp bài chính thức: nếu session đã kết thúc thì coi như thành công và redirect;
        /// ngược lại gọi API SubmitExam để chấm điểm.
        /// </summary>
        public async Task<(bool Success, string? ErrorMessage, long SessionId, bool RedirectToDetails, bool ShowSuccessMessage)> SubmitExamAsync(SubmitExamFormModel model, long userId)
        {
            if (model.ExamSessionId <= 0)
                return (false, "Phiên thi không hợp lệ.", model.ExamSessionId, false, false);

            var (session, error) = await GetOwnedSessionAsync(model.ExamSessionId, userId);
            if (session == null)
                return (false, error, model.ExamSessionId, false, false);

            // Idempotent: nộp lại khi đã kết thúc vẫn redirect thành công, không gọi API lần nữa.
            if (session.Status != (int)ExamSessionStatusEnum.InProgress)
                return (true, null, model.ExamSessionId, true, false);

            var result = await _sessionRepository.SubmitExamAsync(new SubmitExamModel
            {
                ExamSessionId = model.ExamSessionId,
                Answers = BuildSubmitAnswers(model)
            });

            if (!result.Success)
                return (false, result.Message ?? "Nộp bài thất bại.", model.ExamSessionId, false, false);

            return (true, null, model.ExamSessionId, true, true);
        }

        /// <summary>
        /// Hủy bài thi do vi phạm quy chế: gọi API cancel-exam-violation, khóa không cho thi lại.
        /// </summary>
        public async Task<(bool Success, string? ErrorMessage, long SessionId)> CancelDueToViolationAsync(SubmitExamFormModel model, long userId)
        {
            if (model.ExamSessionId <= 0)
                return (false, "Phiên thi không hợp lệ.", model.ExamSessionId);

            var (session, error) = await GetOwnedSessionAsync(model.ExamSessionId, userId);
            if (session == null)
                return (false, error, model.ExamSessionId);

            // Idempotent: đã kết thúc thì coi như thành công
            if (session.Status != (int)ExamSessionStatusEnum.InProgress)
                return (true, null, model.ExamSessionId);

            var result = await _sessionRepository.CancelExamDueToViolationAsync(new SubmitExamModel
            {
                ExamSessionId = model.ExamSessionId,
                ViolationCount = model.ViolationCount,
                Answers = BuildSubmitAnswers(model)
            });

            if (!result.Success)
                return (false, result.Message ?? "Không thể hủy bài thi.", model.ExamSessionId);

            return (true, null, model.ExamSessionId);
        }

        /// <summary>
        /// Chuẩn hóa đáp án từ form trước khi gửi API: loại id không hợp lệ, trùng lặp, và câu chưa chọn đáp án.
        /// </summary>
        private static List<SubmitAnswerModel> BuildSubmitAnswers(SubmitExamFormModel model)
        {
            // Chỉ gửi câu đã chọn ít nhất một đáp án; bỏ id <= 0 và trùng.
            return model.Answers?
                .Select(a => new SubmitAnswerModel
                {
                    QuestionId = a.QuestionId,
                    SelectedAnswerOptionIds = (a.SelectedAnswerOptionIds ?? new List<long>())
                        .Where(id => id > 0)
                        .Distinct()
                        .ToList()
                })
                .Where(a => a.SelectedAnswerOptionIds.Count > 0)
                .ToList() ?? new List<SubmitAnswerModel>();
        }
    }
}

File: lib/services/exam_session_service.dart
import '../core/enums/app_enums.dart';
import '../models/exam/exam_period_assignment.dart';
import '../models/exam/exam_requests.dart';
import '../models/exam/exam_session.dart';
import '../models/exam/exam_set.dart';
import '../repositories/exam_period_assignment_repository.dart';
import '../repositories/exam_session_repository.dart';
import '../repositories/exam_set_repository.dart';
import 'auth_service.dart';

/// Item bài đang làm — gộp phiên thi + phân công + bộ đề.
class InProgressExamItem {
  final ExamSession session;
  final ExamPeriodAssignment? assignment;
  final ExamSet? examSet;

  const InProgressExamItem({
    required this.session,
    this.assignment,
    this.examSet,
  });
}

/// Kết quả bắt đầu thi — tương đương `StartExamServiceResult`.
class StartExamResult {
  final bool success;
  final String? errorMessage;
  final int? sessionId;
  final String? resumeMessage;

  const StartExamResult({
    required this.success,
    this.errorMessage,
    this.sessionId,
    this.resumeMessage,
  });
}

/// Dữ liệu màn làm bài — tương đương `TakeExamViewModel`.
class TakeExamData {
  final ExamSession session;
  final int durationMinutes;
  final DateTime endTime;

  const TakeExamData({
    required this.session,
    required this.durationMinutes,
    required this.endTime,
  });
}

/// Dịch vụ phiên thi — port từ `ExamSessionService.cs`.
class ExamSessionService {
  ExamSessionService(
    this._sessionRepo,
    this._examSetRepo,
    this._assignmentRepo,
  );

  final ExamSessionRepository _sessionRepo;
  final ExamSetRepository _examSetRepo;
  final ExamPeriodAssignmentRepository _assignmentRepo;

  Future<({List<InProgressExamItem> items, int total, String? error})>
      getInProgressList({
    required int userId,
    required int pageIndex,
  }) async {
    final result = await _sessionRepo.getList(
      pageIndex: pageIndex,
      userId: userId,
      status: ExamSessionStatus.inProgress,
    );
    if (!result.success) {
      return (
        items: <InProgressExamItem>[],
        total: 0,
        error: result.message.isNotEmpty
            ? result.message
            : 'Không thể tải dữ liệu từ API.',
      );
    }

    final sessions = result.data ?? [];
    final assignmentsResult = await _assignmentRepo.getMyAssignments();
    final assignmentsById = <int, ExamPeriodAssignment>{};
    final completedAssignmentIds = <int>{};

    if (assignmentsResult.success) {
      for (final a in assignmentsResult.data ?? []) {
        assignmentsById[a.id] = a;
        if (a.status == ExamPeriodAssignmentStatus.completed) {
          completedAssignmentIds.add(a.id);
        }
      }
    }

    final filtered = sessions.where((s) {
      final assignmentId = s.examPeriodAssignmentId;
      return assignmentId == null || !completedAssignmentIds.contains(assignmentId);
    }).toList();

    final examSetCache = <int, ExamSet?>{};
    final items = <InProgressExamItem>[];

    for (final session in filtered) {
      final assignment = session.examPeriodAssignmentId != null
          ? assignmentsById[session.examPeriodAssignmentId]
          : null;
      ExamSet? examSet;
      final examSetId = assignment?.examSetId ?? session.examSetId;
      if (examSetId > 0) {
        if (examSetCache.containsKey(examSetId)) {
          examSet = examSetCache[examSetId];
        } else {
          final examSetResult = await _examSetRepo.getById(examSetId);
          examSet = examSetResult.success ? examSetResult.data : null;
          examSetCache[examSetId] = examSet;
        }
      }
      items.add(
        InProgressExamItem(
          session: session,
          assignment: assignment,
          examSet: examSet,
        ),
      );
    }

    items.sort((a, b) {
      final aStart = a.assignment?.examPeriodStartAt ?? a.session.startedAt;
      final bStart = b.assignment?.examPeriodStartAt ?? b.session.startedAt;
      return aStart.compareTo(bStart);
    });

    return (
      items: items,
      total: items.length,
      error: null,
    );
  }

  Future<({List<ExamSession> items, int total, String? error})> getHistoryList({
    required int userId,
    required int pageIndex,
  }) async {
    final result = await _sessionRepo.getHistory(
      pageIndex: pageIndex,
      userId: userId,
    );
    if (!result.success) {
      return (
        items: <ExamSession>[],
        total: 0,
        error: result.message.isNotEmpty
            ? result.message
            : 'Không thể tải dữ liệu từ API.',
      );
    }
    return (
      items: result.data ?? [],
      total: result.metaData?.totalItems ?? result.data?.length ?? 0,
      error: null,
    );
  }

  Future<ServiceResult<ExamSession>> getOwnedSession(int id, int userId) async {
    final result = await _sessionRepo.getById(id);
    if (!result.success || result.data == null) {
      return ServiceResult.fail(
        result.message.isNotEmpty
            ? result.message
            : 'Không tìm thấy phiên thi.',
      );
    }
    if (userId > 0 && result.data!.userId != userId) {
      return ServiceResult.fail('Bạn không có quyền truy cập phiên thi này.');
    }
    return ServiceResult.ok(result.data);
  }

  Future<StartExamResult> startExam({
    required int examSetId,
    required int examType,
    required int examPeriodAssignmentId,
    required int userId,
  }) async {
    int? existingSessionId;
    final existingResult = await _sessionRepo.getList(
      pageIndex: 1,
      pageSize: 1,
      userId: userId,
      examSetId: examSetId > 0 ? examSetId : null,
      examType: examType,
      status: ExamSessionStatus.inProgress,
    );
    existingSessionId = existingResult.data?.firstOrNull?.id;

    final result = await _sessionRepo.startExam(
      StartExamRequest(
        examPeriodAssignmentId: examPeriodAssignmentId,
        examSetId: examSetId,
        examType: examType,
      ),
    );

    if (!result.success || result.data == null) {
      return StartExamResult(
        success: false,
        errorMessage: result.message.isNotEmpty
            ? result.message
            : 'Không thể bắt đầu bài thi.',
      );
    }

    return StartExamResult(
      success: true,
      sessionId: result.data!.id,
      resumeMessage: existingSessionId != null &&
              existingSessionId == result.data!.id
          ? 'Tiếp tục bài thi đang làm.'
          : null,
    );
  }

  Future<({
    TakeExamData? data,
    String? successMessage,
    String? errorMessage,
    bool redirectToDetails,
  })> getTakeExam(int id, int userId) async {
    final owned = await getOwnedSession(id, userId);
    if (!owned.success || owned.data == null) {
      return (data: null, successMessage: null, errorMessage: owned.errorMessage, redirectToDetails: false);
    }

    final session = owned.data!;
    if (session.status != ExamSessionStatus.inProgress) {
      final msg = switch (session.status) {
        ExamSessionStatus.expired =>
          'Bài thi đã hết giờ và được tính điểm tự động.',
        ExamSessionStatus.cancelled =>
          'Bài thi đã bị hủy do vi phạm quy chế thi.',
        _ => null,
      };
      return (
        data: null,
        successMessage: msg,
        errorMessage: null,
        redirectToDetails: true,
      );
    }

    final examSetResult = await _examSetRepo.getById(session.examSetId);
    final duration = examSetResult.data?.durationMinutes ?? 60;
    final startedAt = session.startedAt.toUtc();
    final endTime = startedAt.add(Duration(minutes: duration));

    return (
      data: TakeExamData(
        session: session,
        durationMinutes: duration,
        endTime: endTime,
      ),
      successMessage: null,
      errorMessage: null,
      redirectToDetails: false,
    );
  }

  Future<ServiceResult<void>> saveProgress({
    required int examSessionId,
    required Map<int, List<int>> answers,
    required int userId,
  }) async {
    final owned = await getOwnedSession(examSessionId, userId);
    if (!owned.success) return ServiceResult.fail(owned.errorMessage!);
    if (owned.data!.status != ExamSessionStatus.inProgress) {
      return ServiceResult.fail('Phiên thi đã kết thúc.');
    }

    final result = await _sessionRepo.saveExamProgress(
      _buildSubmitRequest(examSessionId, answers),
    );
    if (!result.success) {
      return ServiceResult.fail(result.message);
    }
    return ServiceResult.ok();
  }

  Future<ServiceResult<void>> submitExam({
    required int examSessionId,
    required Map<int, List<int>> answers,
    required int userId,
  }) async {
    final owned = await getOwnedSession(examSessionId, userId);
    if (!owned.success) return ServiceResult.fail(owned.errorMessage!);

    if (owned.data!.status != ExamSessionStatus.inProgress) {
      return ServiceResult.ok();
    }

    final result = await _sessionRepo.submitExam(
      _buildSubmitRequest(examSessionId, answers),
    );
    if (!result.success) {
      return ServiceResult.fail(
        result.message.isNotEmpty ? result.message : 'Nộp bài thất bại.',
      );
    }
    return ServiceResult.ok();
  }

  Future<ServiceResult<void>> cancelDueToViolation({
    required int examSessionId,
    required Map<int, List<int>> answers,
    required int violationCount,
    required int userId,
  }) async {
    final owned = await getOwnedSession(examSessionId, userId);
    if (!owned.success) return ServiceResult.fail(owned.errorMessage!);

    if (owned.data!.status != ExamSessionStatus.inProgress) {
      return ServiceResult.ok();
    }

    final request = SubmitExamRequest(
      examSessionId: examSessionId,
      violationCount: violationCount,
      answers: _buildAnswers(answers),
    );

    final result = await _sessionRepo.cancelExamDueToViolation(request);
    if (!result.success) {
      return ServiceResult.fail(
        result.message.isNotEmpty ? result.message : 'Không thể hủy bài thi.',
      );
    }
    return ServiceResult.ok();
  }

  SubmitExamRequest _buildSubmitRequest(
    int examSessionId,
    Map<int, List<int>> answers,
  ) {
    return SubmitExamRequest(
      examSessionId: examSessionId,
      answers: _buildAnswers(answers),
    );
  }

  List<SubmitAnswerRequest> _buildAnswers(Map<int, List<int>> answers) {
    return answers.entries
        .map(
          (e) => SubmitAnswerRequest(
            questionId: e.key,
            selectedAnswerOptionIds: e.value.where((id) => id > 0).toSet().toList(),
          ),
        )
        .where((a) => a.selectedAnswerOptionIds.isNotEmpty)
        .toList();
  }
}

extension _FirstOrNull<E> on Iterable<E> {
  E? get firstOrNull {
    final iterator = this.iterator;
    if (iterator.moveNext()) return iterator.current;
    return null;
  }
}
